### ISSUE TITLE
Provide generic PA plugin that takes textual input for addresses

### ISSUE THREAD

**Issue by Kissaki** [mumble-voip/mumble#1322]
Create a plugin that takes multiple files of some text format (human-editable).

Goal: On game updates, just edit the address offsets - done. No compilation or binary dll.

**Comment by davidebeatrici** [comment]
Should this be integrated in Mumble or a separated plugin?

**Comment by mkrautz** [comment]
It should be a plugin, "text" (name TBD). The plugin loads structured text files describing where to read the offsets.
It could be coded to support multiple games.

Think it of like "link".

**Comment by mkrautz** [comment]
Though I worry that some games will still require some kind of logic to work properly.

I've also thought of doing PA plugins with JavaScript via V4/QJSEngine (http://doc.qt.io/qt-5/qjsengine.html). The only problem with that would be the performance of reading the memory addresses many times per second.

**Comment by davidebeatrici** [comment]
I think that "custom" sounds good as name :smile: 
But how can we communicate to the plugin which text file to load? I was thinking of a GUI for the plugin that can be opened with the "Configuration" button on Mumble and MumblePAHelper.
For the logic it shouldn't be a problem, regarding coordinate system conversions; regarding pointers and multi-modules it can be complicated.

**Comment by hacst** [comment]
Well. It's only a hundred times per second worst case so it's not thaaaaaat performance sensitive. We could always embed LuaJIT if we are concerned about performance there. However allowing the seperation of address and logic updates (the later staying as dll's) would already be quite helpful and doesn't require any scripting. Even arbitrary pointer chain chasing wouldn't require a full blown scripting capability.

**Comment by Kissaki** [comment]
Yeah, after I did some work for this it became clear to me that a pure (address) declarative approach does not work at all. We will need some kind of logic and parser. Maybe using something like flex/bison or the like.

The plugin I started I named "any" as in "any game". IIRC I was also questioning whether a plugin would even be a good place. It looked like using/integrating into Mumbles classes would be more feasable.

**Comment by davidebeatrici** [comment]
We could use files with a dedicated extension, for example ".pa", with the info in them.
Something like this:
> Name: Game
> Description: Game version 1.0
> Game executable: Name.exe
> Coordinate system: Left handed, Y-up
> Avatar = camera: No
> Context support: Yes
> Identity support: Yes
> Avatar position address: 0x100000
> Avatar front address: 0x100010
> Avatar top address: 0x100020
> Camera position address: 0x200000
> Camera front address: 0x200010
> Camera top address: 0x200020
> Context address: 0x400000
> Identity address: 0x500000

**Comment by Kissaki** [comment]
We have plugins that get their specific address from addressspace + X -> resolve/dereference to Y, Y+Z = address

Games also have different coordinate systems, so we need transformations (/flags) for that.

**Comment by Kissaki** [comment]
I don’t think we have a context address ever, apart from Link where we don’t need a plugin anyway.
Context, IIRC, always(?) contains multiple datapoints - which we encode JSON format for usability.

**Comment by davidebeatrici** [comment]
Well, usually the context is the server IP Address + Port or the server ID in a string, a single memory address.

**Comment by hacst** [comment]
If we want to write new plugins in a declarative way that would be quite hard. Keeping existing ones up-2-date would be much simpler (and likely safer).

**Comment by davidebeatrici** [comment]
This plugin would be very useful to quickly test if memory addresses are working as they should. Of course a dedicated up-to-date plugin is much better... we could put a button to send the info to us and then someone will code the plugin.
Thoughts?
